[RFC] Clean up buffer.c build_stl_str_hl variable declarations

war1025 opened the pull request:
I picked this function more or less at random and decided to move the declarations closer to where the variables are used at. 

After further investigation, this is the function that builds the vim statusline.
The code was rather cryptic, so I have fully commented it. The comments may seem excessive, but that is only because after reading the comment it is obvious that the code is doing what the comment says it is. Without the comments it is quite hard to follow.

I have also written a number of tests to cover the basic items supported by statusline. It is by no means a comprehensive test suite for the function.

war1025 committed:
Clean up buffer.c build_stl_str_hl variable declarations

war1025 commented on src/nvim/buffer.c:
```diff
@@ -3062,25 +3040,26 @@ build_stl_str_hl (
-      else {
-        t = (opt == STL_FULLPATH) ? wp->w_buffer->b_ffname
-            : wp->w_buffer->b_fname;
-        home_replace(wp->w_buffer, t, NameBuff, MAXPATHL, TRUE);
```
the linter told me to change `TRUE` to `true` on this line, but I'm not sure why it complained because I didn't change this line and it was fine before?

ZyX-I replied:
@war1025 Linter silences the errors for *blocks* containing *three* lines, line with the error in the middle. Should you change any of the lines in the block you need to also change adjacent lines.

justinmk commented:
Starting with https://github.com/neovim/neovim/pull/3297 we would like prefer that cleanup PRs include a couple unit tests that cover the function being changed, so we can feel more confident about regressions and future changes / vim patch merges. 

The unit tests _should_ initialize all globals so it won't be as painful as it used to be. 

https://coveralls.io/builds/3555525/source?filename=src%2Fnvim%2Fbuffer.c indicates that `build_stl_str_hl` is not covered by our tests.

war1025 committed:
Add comments to build_stl_str_hl

war1025 committed:
Rename build_stl_str_hl input and output buffer pointers

war1025 committed:
Add out_end_p to mark last valid character in statusline buffer

war1025 committed:
Fix lint

war1025 committed:
Add documentation about what TABPAGENR and TABCLOSENR do

war1025 commented:
@justinmk I have this pull pretty well cleaned up. I am looking into writing tests, but I need some info on how to access some things:

I need access to a `win_T*`. `curwin` would work fine for me.
I need to allocate a `char_u*` output buffer. I was thinking maybe `toCstr()` would create one for me if I gave it a string?

I think the other values I can figure out. There are a couple structs at the end of the arguments list, but maybe I can ignore those for now.

war1025 committed:
Fix shadowed variable in build_stl_str_hl

war1025 committed:
Start adding unit tests

justinmk commented:
>  need access to a win_T*. curwin would work fine for me.

`curwin` is fine.

> I need to allocate a char_u* output buffer. I was thinking maybe toCstr() would create one for me if I gave it a string?

Yes, that should work.

war1025 commented:
Attempted to write unit tests for this function. Not enough things are initialized for it to work in anything but trivial cases. Will try to figure out how the functional tests stuff works.

war1025 commented:
Functional tests are pretty slick...

war1025 commented:
I have done all the work I plan to do on this pull request. I believe someone still needs to go through it before I can mark it was `[RDY]`. I will check back periodically to see if there has been any activity.

ZyX-I commented on src/nvim/buffer.c:
```diff
@@ -2785,8 +2785,8 @@ void free_titles(void)
-int 
-build_stl_str_hl (
+int
+build_stl_str_hl(
```
I think that since you already touch this you should use style from the style guide:

```C
int build_stl_str_hl(
```

ZyX-I commented on src/nvim/buffer.c:
```diff
@@ -2833,15 +2816,16 @@ build_stl_str_hl (
+     DECIMAL     = 10,
+     OCTAL       = 8,
+     HEXIDECIMAL = 16
+  } number_base;
```
It is very strange to see `typedef` inside the function. Also note that according to the style guide this needs to be

```C
typedef enum {
    kNumBaseDecimal = 10,
    kNumBaseOctal = 8,
    kNumBaseHexadecimal = 16,
} NumberBase;
```
```

war1025 replied:
Do you have a better solution than the enum?

I am fine updating to match the style guide, but if you have an alternate idea I'd also be happy to change it.

ZyX-I replied:
@war1025 Why do you need better solution? Just do not place it *inside the function* and correct naming.

war1025 replied:
I see, your issue was with the enum being inside the function. That is the piece that I missed. I will update.

war1025 committed:
Updates from review

ZyX-I commented on src/nvim/buffer.c:
```diff
@@ -2770,6 +2770,16 @@ void free_titles(void)
+/**
+ * Enumeration specifying the valid numeric bases that can
+ * be used when printing numbers in the status line.
+ **/
```
Style guide accepts *only* `//`-style comments (specifically `///` for doxygen). `/*…*/` would be fine only inside a macros and only as long as they are one-line.

war1025 committed:
Switch comments to match style guide

war1025 commented:
@ZyX-I I updated the comments in the function to match the style guide I believe.

war1025 commented:
ping @justinmk @ZyX-I 

I put a good deal of time into this pull. I'd like to see it merged rather than just sitting in review indefinitely. Is there anything I can do to speed the process along?

justinmk commented:
It won't be indefinite. I will take a look soon.

war1025 commented:
Thanks

justinmk commented on src/nvim/buffer.c:
```diff
@@ -2770,58 +2770,59 @@ void free_titles(void)
   struct stl_item {
+    // Where the item starts in the status line output buffer
     char_u          *start;
+    // The minimum width of the item (at most 50 characters)
```
better to leave out potentially stale info such as "at most 50"

war1025 replied:
__Done__

justinmk commented on src/nvim/buffer.c:
```diff
@@ -2855,98 +2849,152 @@ build_stl_str_hl (
-     */
-    s++;
-    if (*s == NUL)      /* ignore trailing % */
+    // The rest of this loop wil handle a single `%` item.
```
typo

war1025 replied:
__Done__

justinmk commented on src/nvim/buffer.c:
```diff
@@ -2855,98 +2849,152 @@ build_stl_str_hl (
+      char_u *t = item[groupitem[groupdepth]].start;
+      *out_p = NUL;
+      long l = vim_strsize(t);
+      // }
```
Avoid "comment scopes" in most cases. E.g., this comment scope is already fairly clear because of the blank line below. In other cases, a comment on a `while()` is already "scoped" by the `while(){...}` scope.

justinmk commented on src/nvim/buffer.c:
```diff
@@ -2954,76 +3002,129 @@ build_stl_str_hl (
-      l = -1;
+
+    // Denotes that the item should be left-aligned.
+    // This is tracked by using a negative length.
```
is the "negative length" comment still true? Seems like it is tracked via `left_align` boolean now.

justinmk commented on src/nvim/buffer.c:
```diff
@@ -3272,212 +3433,362 @@ build_stl_str_hl (
+    }
 
-    if (opt == STL_VIM_EXPR)
+    // Free the string buffer if we allocated it.
```
this might be a bit obvious :}

war1025 replied:
I changed the wording a little. 
I mainly want to leave this comment in because the `if()` statement is checking the one place where we did allocate the string buffer. The comment makes it more obvious why we are doing the deallocation only in that case (or potentially other cases in the future).

justinmk commented on src/nvim/buffer.c:
```diff
@@ -3272,212 +3433,362 @@ build_stl_str_hl (
 
-  if (usefmt != fmt)
+  // Null terminate the output buffer
+  *out_p = NUL;
```
comment is probably not needed, this is a common idiom in Vim codebase (and C in general I guess)

war1025 replied:
I removed this comment and added one a couple lines down stating more clearly the point I was trying to get across.

```
  // We have now processed the entire statusline format string.
  // What follows is post-processing to handle alignment and
  // highlighting factors.
```

justinmk commented on src/nvim/buffer.c:
```diff
@@ -3272,212 +3433,362 @@ build_stl_str_hl (
-      for (l = 0; l < itemcnt; l++)
-        if (item[l].start > s)
+        // Fill up for half a double-wide character.
+        // XXX : This seems impossible given the exit condition above?
```
Could have been left-over from Vim source before removing `#ifdef` blocks which maybe confused the original author(s). Would be worth peeking at the Vim source here.

war1025 replied:
I looked at the source. Appears just the same there. I'm going to remove it since as far as I can tell it doesn't do anything.

justinmk commented on src/nvim/buffer.c:
```diff
@@ -3272,212 +3433,362 @@ build_stl_str_hl (
+      *trunc_p = '<';
+
+      // Advance the pointer to the end of the string
+      trunc_p = trunc_p + STRLEN(trunc_p);
```
why is it safe to move this out of the `while()` loop below?

war1025 replied:
In the `while()` loop, the last step is to set `*trunc_p = NUL`, which means on every round except the first, `STRLEN(trunc_p) == 0`, correct?

justinmk replied:
Correct me if I'm wrong: on each loop, `*trunc_p++ = fillchar;` overwrites the previous `NUL`, then the pointer is incremented and its value set to `NUL`. That means `STRLEN(trunc_p)` would be longer than it started with.

war1025 replied:
The string is getting longer after each loop, however we are also moving `trunc_p` a character ahead each loop.

`STRLEN(trunc_p)` counts the number of characters until a null is found starting from `trunc_p`, which at the end of the loop we just set to `NUL`. So `STRLEN(trunc_p)` has to be `0` on each subsequent loop.

oni-link replied:
This line also needs to check `width+1<maxwidth` before changing `trunc_p`. Otherwise too long tablines could crash nvim.

justinmk replied:
Thanks @oni-link . @war1025 can you fix that?

justinmk commented on src/nvim/buffer.c:
```diff
@@ -2855,98 +2849,152 @@ build_stl_str_hl (
-      if (groupdepth > 0)
+
+    // STL_MIDDLEMARK denotes the separation place between
+    // left and right aligned items.
```
"denotes the separation place" => "separates". Then this comment fits on 1 line.

justinmk commented on src/nvim/buffer.c:
```diff
@@ -2855,98 +2849,152 @@ build_stl_str_hl (
-      l = vim_strsize(t);
+      // { Determine how long the group is.
+      //   Note: We set the current output position to null
+      //         so `vim_strsize` will work.
```
this could be inline to the right of `*out_p = NUL;`

war1025 replied:
__Done__ I removed the scoping. Inline comments are a pet-peeve of mine. So I chose not to make that change.

justinmk commented on src/nvim/buffer.c:
```diff
@@ -3088,42 +3205,62 @@ build_stl_str_hl (
-      o_curbuf = curbuf;
-      o_curwin = curwin;
+      // Switch the curbuf and curwin to the buffer and window we are
+      // evaluating the statusline for.
```
this is extremely common idiom in Vim source, so should probably leave out this comment.

war1025 replied:
I am a bit on the fence about this one. One of the big issues I have with open source projects vs the codebase where I work is that everything tends to be super under-documented. 

If I am random person wondering how the code works, I have to know a bunch of quirks about the codebase before I can understand anything. 

IMHO if you are changing a global variable to something that isn't what it actually is, that deserves a comment, no matter how common it is. If we don't want the comment, we should refactor things to not rely on the global variables.

That said, if you really want to, I will remove this comment.

war1025 replied:
There is a `WITH_BUFFER` macro. It looks like similar commands exist to create a `WITH_WINDOW` macro. I would not be opposed to creating that and using it here.

justinmk commented:
@war1025 Thanks for the thoughtful and informed changes. 

Other than my 1 technical question and some nudges in the direction of "brevity without sacrificing clarity", it looks good to me. Let me know if you can address the comments and rebase.

war1025 committed:
Cleanup from review

war1025 commented:
I went through and made most of the changes you suggested. Please make another pass when you have time.

war1025 commented:
@justinmk I added two macros `WITH_WINDOW` and `WITH_WINDOW_IN_TAB` to hide the switching of `curbuf`, `curwin`, and `curtab`. I updated the places in the codebase that used the `switch_win()` function. I imagine there are many more places that could be updated later.

Please let me know what you think of this change.

justinmk commented:
@war1025 I really like this approach. I _was_ going to off-handedly suggest using a function or macro in a later PR... this is better than expected :) In either case it could break if there is an early `return`--but at least with this approach the scope is explicit, and could be guarded with a linter...

justinmk commented on src/nvim/buffer.c:
```diff
@@ -3107,42 +3222,54 @@ build_stl_str_hl (
 
+      // Store the current buffer number as a string variable
       vim_snprintf((char *)tmp, sizeof(tmp), "%d", curbuf->b_fnum);
       set_internal_string_var((char_u *)"actual_curbuf", tmp);
```
Amusing side-note: this is a secret "reserved" variable name. In N/Vim, try

    :let g:actual_curbuf='foo' 

with a statusline like:

    :set statusline=%{bufnr('')}

in N/Vim--it will be immediately unlet.

ZyX-I replied:
It is not “secret”: do `helpg actual_curbuf`, you will see a reference under `:h 'statusline'`. But it is missing a tag.

justinmk commented on src/nvim/buffer.c:
```diff
@@ -3107,42 +3222,54 @@ build_stl_str_hl (
+      WITH_WINDOW(wp, {
+        // Note: The result stored in `t` is unused.
+        str = eval_to_string_safe(out_p, &t, use_sandbox);
+      });
```
Sadly, this is not equivalent. `switch_win()` calls `block_autocmds()` which I am assuming is the cause of some weird behavior with this PR. Test case:

```vim
    autocmd! TermOpen * echom 'termopen' localtime()
    function! Foo()
      doautocmd TermOpen
    endfunction
    set statusline=%{Foo()}
```

The autocmd is not triggered by statusline evaluation ("No matching autocmds" is reported every time statusline redraws).

Perhaps `WITH_WINDOW` could take a parameter which prevents disabling of autocmds?

war1025 commented:
@justinmk I have pulled the `WITH_WINDOW` thing out since that looks like it will be a little more work than I initially expected. I removed the comment that started the discussion.

Let me know if there is anything else you will like changed.

justinmk commented:
@war1025  did you see 
https://github.com/neovim/neovim/pull/3339#discussion_r44507563

war1025 commented:
@justinmk I had not seen that comment. Responded.
